app/components/file_attachment_list/file_attachment_document.js

```diff
@@ -0,0 +1,292 @@
+    'application/vnd.openxmlformats-officedocument.presentationml.presentation',
+    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
+    'application/xml',
+    'text/csv'
+];
+
+export default class FileAttachmentDocument extends PureComponent {
+    static propTypes = {
+        iconHeight: PropTypes.number,
+        iconWidth: PropTypes.number,
+        file: PropTypes.object.isRequired,
+        theme: PropTypes.object.isRequired,
+        wrapperHeight: PropTypes.number,
+        wrapperWidth: PropTypes.number
+    };
+
+    static defaultProps = {
+        iconHeight: 65,
+        iconWidth: 65,
+        wrapperHeight: 100,
+        wrapperWidth: 100
+    };
+
+    static contextTypes = {
+        intl: intlShape
+    };
+
+    state = {
+        didCancel: false,
+        downloading: false,
+        progress: 0
+    };
+
+    componentDidMount() {
+        this.mounted = true;
+    }
+
+    componentWillUnmount() {
+        this.mounted = false;
+    }
+
+    cancelDownload = () => {
+        if (this.mounted) {
+            this.setState({didCancel: true});
+        }
+
+        if (this.downloadTask) {
+            this.downloadTask.cancel();
+        }
+    };
+
+    downloadAndPreviewFile = async (file) => {
+        const path = `${DOCUMENTS_PATH}/${file.name}`;
+
+        this.setState({didCancel: false});
+
+        try {
+            const isDir = await RNFetchBlob.fs.isDir(DOCUMENTS_PATH);
+            if (!isDir) {
+                try {
+                    await RNFetchBlob.fs.mkdir(DOCUMENTS_PATH);
+                } catch (error) {
+                    // Do nothing
+                }
+            }
+
+            const options = {
+                session: file.id,
+                timeout: 10000,
+                indicator: true,
+                overwrite: true,
+                path
+            };
+
+            const exist = await RNFetchBlob.fs.exists(path);
+            if (exist) {
+                this.openDocument(file, 0);
+            } else {
+                this.setState({downloading: true});
+                this.downloadTask = RNFetchBlob.config(options).fetch('GET', getFileUrl(file.id)).
+                    progress((received, total) => {
+                        const progress = (received / total) * 100;
+                        if (this.mounted) {
+                            this.setState({progress});
+                        }
+                    });
+
+                await this.downloadTask;
+                if (this.mounted) {
+                    this.setState({
+                        progress: 100
+                    }, () => {
+                        // need to wait a bit for the progress circle UI to update to the give progress
+                        this.openDocument(file);
+                    });
+                }
+            }
+        } catch (error) {
+            RNFetchBlob.fs.unlink(path);
```

Comment:
What happens with this if the file doesn't exist? Does this error out?

Reply:
it does not error out, it just doesn't do anything